drivers/lora/sx12xx_common.c

```diff
@@ -80,14 +94,25 @@ int sx12xx_lora_recv(const struct device *dev, uint8_t *data, uint8_t size,
 {
 	int ret;
 
+	/* Ensure available, decremented by sx12xx_ev_rx_done or on timeout */
+	if (!MODEM_CLAIM(dev_data)) {
+		return -EBUSY;
+	}
+
+	/* Clear any previously triggered packets */
+	k_sem_take(&dev_data.data_sem, K_NO_WAIT);
+
 	Radio.SetMaxPayloadLength(MODEM_LORA, 255);
 	Radio.Rx(0);
 
 	ret = k_sem_take(&dev_data.data_sem, timeout);
 	if (ret < 0) {
 		LOG_INF("Receive timeout");
 		/* Manually transition to sleep mode on timeout */
-		Radio.Sleep();
+		if (MODEM_FREE(dev_data)) {
```

Comment:
I'm not sure I understand why you need to check if you can unlock the mutex before entering the sleep state. That seems odd since you know you had exclusive access to the radio until the point you released it.

Reply:
Because the receive timeout can collide with a packet actually being received, resulting in both code paths trying to call `Radio.Sleep()`, which we can't allow.

Reply:
That sounds like a race condition in that case. I think there are cases where the data semaphore, at least in the subsequent changes, can end up in an unexpected state if there is an unfortunate sequence of receive, a timeout, and a send operation. The receive race in this change is _probably_ benign, but I don't like the idea of having code that is racy. It's going to come back and bite us in the rear at some point.

Reply:
It's racy by definition because we are waiting a fixed duration for an interrupt from an external device. The interrupt can always arrive after we have decided we have timed out, but before we take any action based on that decision. The next update should handle this properly.